sources/api/pluto/src/main.rs

```diff
@@ -270,11 +271,76 @@ fn pause_container_uri(region: &str) -> Result<String> {
     ))
 }
 
+/// Calculates the amount of memory to reserve for kubeReserved in mebibytes.
+/// This function is taken from:
+/// https://github.com/awslabs/amazon-eks-ami/blob/db28da15d2b696bc08ac3aacc9675694f4a69933/files/bootstrap.sh#L228-L239
+fn get_memory_mebibytes_to_reserve(client: &Client, session_token: &str) -> Result<String> {
+    let max_num_pods = get_max_pods(&client, &session_token)?;
+    let mebibytes_unit = "Mi";
+    let max_num_pods_u32 = max_num_pods.parse::<u32>().context(error::ParseToU32 {
+        object: max_num_pods,
+    })?;
+    let memory_to_reserve = max_num_pods_u32 * 11 + 255;
+
+    Ok(memory_to_reserve.to_string() + mebibytes_unit)
+}
+
+/// Calculates the amount of CPU to reserve for kubeReserved in millicores
+/// from the total number of vCPUs available on the instance.
+/// We are using these CPU ranges from GKE
+/// (https://cloud.google.com/kubernetes-engine/docs/concepts/cluster-architecture#node_allocatable):
+///  6% of the first core
+///  1% of the next core (up to 2 cores)
+///  0.5% of the next 2 cores (up to 4 cores)
+///  0.25% of any cores above 4 cores
+/// This function is taken from:
+/// https://github.com/awslabs/amazon-eks-ami/blob/db28da15d2b696bc08ac3aacc9675694f4a69933/files/bootstrap.sh#L241-L251
+fn get_cpu_millicores_to_reserve() -> Result<String> {
```

Comment:
These functions: `get_cpu_millicores_to_reserve` and `get_resource_to_reserve_in_range` are kind of hard to follow. I realize we took them from elsewhere, but is it just me, or does the algorithm described in the function's documentation seem simpler than what the function is doing?

The description sounds like this (possibly but with constants for readability)

```rust
match num_cores {
    1 => 600,
    2 => 600 + 100,
    3 => 600 + 100 + 50,
    4 => 600 + 100 + 50 + 50,
    _ => 600 + 100 + 50 + 50 + ((num_cores - 4) * 25),
}
```

Maybe I just don't get it, and maybe that's OK.

Reply:
Yeah. I think your idea is more readable. But I want to make it consistent with `get_cpu_millicores_to_reserve ` and `get_resource_to_reserve_in_range ` in amazon-eks-ami, where I got the idea from.  What do you think? still keep it same as in amazon-eks-ami or make it more readable?

Reply:
I think it's fine to keep the logic that we have copied from elsewhere. However I would like to see it tested. There's one small change we can make to this function that will make it testable.

Instead of calling `num_cpus::get()` *inside* this function, move it *outside* this function and take num_cpus as an input argument:

```rust
fn get_cpu_millicores_to_reserve(num_cpus: usize) -> Result<String> {
  // ...
}
```

Now write tests for this function given the following inputs
0, 1, 2, 3, 4, 5 and 6.

Reply:
Yeah. Definitely!

Reply:
I see your tests and it seems to suggest that my interpretation of the algo was correct.

By the way, I actually prefer representing test cases independently, like you did when they looked like this:

```rust
#[test]
    fn get_cpu_millicores_to_reserve_0_ok() {
        let num_cpus: usize = 0;
        assert_eq!(get_cpu_millicores_to_reserve(num_cpus).unwrap(), "0m");
    }
```

Instead of building a map or a vec and looping through test cases. (I realize that test case loops are the Go way, but I never liked it). My reasoning is
- In a loop, you don't run the test cases after the one that failed. Sometimes seeing the complete PASS/FAIL picture is helpful.
- When debugging, you have to comment out previous test cases if you're trying to isolate the failure case.